### PR TITLE
better perfs., more reliable astro builds, use SharedWorkers, etc...

### PR DESCRIPTION
**Pull request by okikio**
@natemoo-re Here is the Changelog:
> _**Note**: more testing is required, as well as better UI/UX for loading previews, as well a system for isolating the preview from the rest of the page._

* Fixed #13 
* Created the `WebWorker.ts` file, it allows for use of `SharedWorkers` with the same API surface as normal workers, meaning you can turn Monaco workers into SharedWorkers.
* Removed the `esbuild.ts`, `astro.ts`, and `html.ts` workers, and replaced them with a much bulkier `build.ts` worker, this allows for faster compile. Also, update `editor.ts` & `typescript.ts`.
* Add hover over imports for more info. to Monaco
* Use `@okikio/emitter` for reliable event emitting, and worker/main thread communication
* Update `App.tsx` to work with the new worker updates
* Enable minimap for Monaco
* Update esbuild plugins
* etc...

https://user-images.githubusercontent.com/17222836/136842826-0e2ea037-7f7c-44d0-8241-47695ddcfc06.mp4

### CODE REVIEW

File: src/editor/plugins/cdn.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+// Based on https://github.com/okikio/bundle/blob/main/src/ts/plugins/cdn.ts
```
Comment (natemoo-re) [inline src/editor/plugins/cdn.ts:1]: Thanks for linking out! 🙌 
You deserve all the credit for these!

File: src/editor/plugins/json.ts [hunk 2]
```diff
@@ -14,14 +14,14 @@ export const JSON_PLUGIN = (): Plugin => {
                 const { content, url } = await fetchPkg(pathUrl);
 
                 return {
-                    contents: `
-                    export const value = ${content};
-                    export default value;`,
+                    contents: `\
```
Comment (natemoo-re) [inline src/editor/plugins/json.ts:17]: Typo with this leading `slash`?

File: src/index.ts [hunk 3]
```diff
@@ -1,25 +1,26 @@
...
+// });
 
-const htmlWorker = new Worker(HTML_WORKER_URL, {
-  name: 'html-worker',
-  type: 'module',
-});
+// const htmlWorker = new Worker(HTML_WORKER_URL, {
+//   name: 'html-worker',
+//   type: 'module',
+// });
```
Comment (natemoo-re) [inline src/index.ts:23]: Feel free to drop these!

File: src/index.ts [hunk 4]
```diff
@@ -33,5 +34,5 @@ const htmlWorker = new Worker(HTML_WORKER_URL, {
       console.error(`Could not load share URL`);
     }
   }
-  render({ Monaco, esbuildWorker, astroWorker, htmlWorker, initialModels });
+  render({ Monaco, /* esbuildWorker, astroWorker, htmlWorker, */ initialModels });
```
Comment (natemoo-re) [inline src/index.ts:37]: Same here, remove comment

File: src/utils/WebWorker.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
+// Based on https://github.com/okikio/bundle/blob/main/src/ts/util/WebWorker.ts
+import { EventEmitter } from '@okikio/emitter';
+export class WebWorker implements EventTarget, AbstractWorker {
+    static SharedWorkerSupported = "SharedWorker" in globalThis;
+    ActualWorker: SharedWorker | Worker;
```
Comment (natemoo-re) [inline src/utils/WebWorker.ts:5]: This is really cool!

**Review by natemoo-re — APPROVED** [review]
Looks amazing! A few small things but otherwise looks great. I can fix some of my nits if you don't get to them.

File: src/editor/plugins/json.ts (hunk 2 above)
Comment (natemoo-re) [reply src/editor/plugins/json.ts:17]: This is the only one I'm concerned about before merging, @okikio!

File: src/utils/WebWorker.ts (hunk 5 above)
Comment (natemoo-re) [reply src/utils/WebWorker.ts:5]: Nice! If you want to open another PR and pull in your lib, feel free.

**Comment by okikio** [comment]
@natemoo-re I've resolved all issues, it should be a go for merge

**Pull request merged**